node/src/utils/round_robin.rs

```diff
@@ -31,12 +33,44 @@ pub struct WeightedRoundRobin<I, K> {
     slots: Vec<Slot<K>>,
 
     /// Actual queues.
-    queues: HashMap<K, Mutex<VecDeque<I>>>,
+    queues: HashMap<K, QueueState<I>>,
 
     /// Number of items in all queues combined.
     total: Semaphore,
 }
 
+/// State that wraps queue and its event count.
+#[derive(Debug)]
+struct QueueState<I> {
+    event_count: AtomicUsize,
+    queue: Mutex<VecDeque<I>>,
+}
+
+impl<I> QueueState<I> {
+    fn new() -> Self {
+        QueueState {
+            event_count: AtomicUsize::new(0),
+            queue: Mutex::new(VecDeque::new()),
+        }
+    }
+
+    #[inline]
+    async fn push_back(&self, element: I) {
+        self.queue.lock().await.push_back(element);
+        self.event_count.fetch_add(1, Ordering::SeqCst);
+    }
+
+    #[inline]
+    fn dec_count(&self) {
```

Comment:
Hmm, we have a combined function for pushing elements and increasing the count, but a separate one for decreasing the count.

Should this not instead be a `pop` function?

Reply:
We need the separate for decreasing b/c of the way that `pop()/crank()` is implemented – it first takes the lock on the queue, then checks if it's empty and if not tries to `pop()` from the queue itself. If we have just one method for `pop()` then the logic in the `round_robin.rs` has to be changed. I had one at the beginning but it was deadlocking as I explained to you earlier.

Reply:
Hmm, let's go for the change that we know is correct then, instead of overcorrecting this. Thanks :+1: